### ISSUE TITLE
failed:[ShapeInferenceError] Attribute pads has incorrect size, when using onnxruntime to infer onnx model converted from pytorch model

### ISSUE THREAD

**Issue by xybai** [microsoft/onnxruntime#1021]
Hi, I use torch.onnx to convert a pytorch model to onnx model and use onnxruntime to use it to infer data (no gpu). The following error happends when running "onnxsession = onnxruntime.InferenceSession(args.onnxmodel)".
![image](https://user-images.githubusercontent.com/10529016/57667476-3752de00-7636-11e9-81ec-7247c8a40f5b.png)

Hope you can have a check. Thanks!

- OS Platform and Distribution: windows
- ONNX Runtime installed from (source or binary): anaconda
- ONNX Runtime version: 0.4.0
- Python version: 3.6.4

The conversion code is 

        input_ids_query = torch.zeros([args.eval_batch_size, 15], dtype=torch.long)
        input_ids_passage = torch.zeros([args.eval_batch_size, 70], dtype=torch.long)
        torch.onnx.export(model, (input_ids_query, input_ids_passage), "model.onnx",
                          input_names=["input_ids_query", "input_ids_passage"],
                          output_names=["score_logits"],
                          aten=False,
                          operator_export_type=torch.onnx.OperatorExportTypes.ONNX,
                          verbose=True,
                          _retain_param_name=True,
                          opset_version=9,
                          do_constant_folding=True)

The infer code is 

onnxsession = onnxruntime.InferenceSession(args.onnxmodel)
            inputs = {"input_ids_query": input_ids_query.numpy(), "input_ids_passage": input_ids_passage.numpy()}
            outputs = ["score_logits"]
            score_logits = onnxsession.run(outputs, inputs)[0]

**Comment by hariharans29** [comment]
Hi @xybai,

Could you please share the model for further investigation ? 

Thanks!

**Comment by xybai** [comment]
> Hi @xybai,
> 
> Could you please share the model for further investigation ?
> 
> Thanks!

Hi, this is my model skeleton. Please have a look.

class TestModel(nn.Module):
    def __init__(self):
        super(TestModel, self).__init__()
        self.word_embedding = nn.Embedding(num_embeddings=10000, embedding_dim=100)
        
        self.queryRNN = nn.GRU(input_size=100, hidden_size=256, num_layers=2, batch_first=True,
                               dropout=0.5, bidirectional=False)
        self.passageRNN = nn.GRU(input_size=100, hidden_size=256, num_layers=2, batch_first=True,
                                  dropout=0.5, bidirectional=False)
        
        self.queryMatchAtten = MatchAttention()
        self.passageMatchAtten = MatchAttention()
        self.queryMaxPooling = nn.MaxPool2d(kernel_size=(10, 1), stride=1)
        self.passageMaxPooling = nn.MaxPool2d(kernel_size=(50, 1), stride=1)
        
        self.mlp1 = nn.Linear(in_features=768, out_features=512)
        self.batchNorm = nn.BatchNorm1d(512)
        self.relu = nn.ReLU()
        self.mlp2 = nn.Linear(in_features=512, out_features=2)
        self.score_mlp2 = nn.Linear(in_features=512, out_features=1)

    def forward(self, input_ids_query, input_ids_passage, labels=None, scores=None):
        word_embeddings_query = self.word_embedding(input_ids_query)
        word_embeddings_passage = self.word_embedding(input_ids_passage)
        
        query_vector, query_h_n = self.queryRNN(word_embeddings_query)
        passage_vector, passage_h_n = self.passageRNN(word_embeddings_passage)

        query_atten_vector = self.queryMatchAtten(query_vector, passage_vector)
        passage_atten_vector = self.passageMatchAtten(passage_vector, query_vector)

        final_query_vector = self.queryMaxPooling(query_atten_vector)
        final_passage_vector = self.passageMaxPooling(passage_atten_vector)

        final_query_vector = final_query_vector[:,-1,:]
        final_passage_vector = final_passage_vector[:,-1,:]

        mul_vector = torch.mul(final_query_vector, final_passage_vector)
        tmp = torch.cat((final_query_vector, mul_vector), -1)
        query_passage_vector = torch.cat((tmp, final_passage_vector), -1)

        mlp_vector1 = self.mlp1(query_passage_vector)
        batch_norm_vector = self.batchNorm(mlp_vector1)
        relu_vector = self.relu(batch_norm_vector)

        tmp_logits = self.score_mlp2(relu_vector)
        score_logits = torch.sigmoid(tmp_logits)
        return score_logits

**Comment by hariharans29** [comment]
> > Hi @xybai,
> > Could you please share the model for further investigation ?
> > Thanks!
> 
> Hi, this is my model skeleton. Please have a look. Thanks!
> 
> ```
> class MatchAttention(nn.Module):
>     def __init__(self):
>         super(MatchAttention, self).__init__()
>         self.softmax = nn.Softmax(dim=-1)
> 
>     def forward(self, x, y):
>         multi = x.bmm(y.transpose(2, 1))
>         weight = self.softmax(multi)
>         x_atten = weight.bmm(y)
>         return x_atten
> 
> 
> class TestModel(nn.Module):
>     def __init__(self):
>         super(TestModel, self).__init__()
>         self.word_embedding = nn.Embedding(num_embeddings=10000, embedding_dim=100)
> 
>         self.queryRNN = nn.GRU(input_size=100, hidden_size=256, num_layers=2, batch_first=True,
>                                dropout=0.5, bidirectional=False)
>         self.passageRNN = nn.GRU(input_size=100, hidden_size=256, num_layers=2, batch_first=True,
>                                   dropout=0.5, bidirectional=False)
> 
>         self.queryMatchAtten = MatchAttention()
>         self.passageMatchAtten = MatchAttention()
>         self.queryMaxPooling = nn.MaxPool2d(kernel_size=(15, 1), stride=1)
>         self.passageMaxPooling = nn.MaxPool2d(kernel_size=(70, 1), stride=1)
> 
>         self.mlp1 = nn.Linear(in_features=768, out_features=512)
>         self.batchNorm = nn.BatchNorm1d(512)
>         self.relu = nn.ReLU()
>         self.mlp2 = nn.Linear(in_features=512, out_features=2)
>         self.score_mlp2 = nn.Linear(in_features=512, out_features=1)
> 
>     def forward(self, input_ids_query, input_ids_passage, labels=None, scores=None):
>         word_embeddings_query = self.word_embedding(input_ids_query)
>         word_embeddings_passage = self.word_embedding(input_ids_passage)
> 
>         query_vector, query_h_n = self.queryRNN(word_embeddings_query)
>         passage_vector, passage_h_n = self.passageRNN(word_embeddings_passage)
> 
>         query_atten_vector = self.queryMatchAtten(query_vector, passage_vector)
>         passage_atten_vector = self.passageMatchAtten(passage_vector, query_vector)
> 
>         final_query_vector = self.queryMaxPooling(query_atten_vector)
>         final_passage_vector = self.passageMaxPooling(passage_atten_vector)
> 
>         final_query_vector = final_query_vector[:,-1,:]
>         final_passage_vector = final_passage_vector[:,-1,:]
> 
>         mul_vector = torch.mul(final_query_vector, final_passage_vector)
>         tmp = torch.cat((final_query_vector, mul_vector), -1)
>         query_passage_vector = torch.cat((tmp, final_passage_vector), -1)
> 
>         mlp_vector1 = self.mlp1(query_passage_vector)
>         batch_norm_vector = self.batchNorm(mlp_vector1)
>         relu_vector = self.relu(batch_norm_vector)
> 
>         tmp_logits = self.score_mlp2(relu_vector)
>         score_logits = torch.sigmoid(tmp_logits)
>         return score_logits
> ```

Hi @xybai,

Thanks for the reply.

This is great, but I do need the .onnx file. This will help me to

1) Run it through the [onnx model checker](https://github.com/onnx/onnx/blob/master/docs/PythonAPIOverview.md) to determine if it is a valid ONNX model. There could be an exporter bug that could have created an invalid ONNX model which needs to be tracked differently.

2) Use the model to actually to score a dummy tensor and debug issues along the way.

I am not sure the model skeleton will help me do the above. So, if you can share your model, that would be great!

Thanks!

**Comment by xybai** [comment]
> > > Hi @xybai,
> > > Could you please share the model for further investigation ?
> > > Thanks!
> > 
> > 
> > Hi, this is my model skeleton. Please have a look. Thanks!
> > ```
> > class MatchAttention(nn.Module):
> >     def __init__(self):
> >         super(MatchAttention, self).__init__()
> >         self.softmax = nn.Softmax(dim=-1)
> > 
> >     def forward(self, x, y):
> >         multi = x.bmm(y.transpose(2, 1))
> >         weight = self.softmax(multi)
> >         x_atten = weight.bmm(y)
> >         return x_atten
> > 
> > 
> > class TestModel(nn.Module):
> >     def __init__(self):
> >         super(TestModel, self).__init__()
> >         self.word_embedding = nn.Embedding(num_embeddings=10000, embedding_dim=100)
> > 
> >         self.queryRNN = nn.GRU(input_size=100, hidden_size=256, num_layers=2, batch_first=True,
> >                                dropout=0.5, bidirectional=False)
> >         self.passageRNN = nn.GRU(input_size=100, hidden_size=256, num_layers=2, batch_first=True,
> >                                   dropout=0.5, bidirectional=False)
> > 
> >         self.queryMatchAtten = MatchAttention()
> >         self.passageMatchAtten = MatchAttention()
> >         self.queryMaxPooling = nn.MaxPool2d(kernel_size=(15, 1), stride=1)
> >         self.passageMaxPooling = nn.MaxPool2d(kernel_size=(70, 1), stride=1)
> > 
> >         self.mlp1 = nn.Linear(in_features=768, out_features=512)
> >         self.batchNorm = nn.BatchNorm1d(512)
> >         self.relu = nn.ReLU()
> >         self.mlp2 = nn.Linear(in_features=512, out_features=2)
> >         self.score_mlp2 = nn.Linear(in_features=512, out_features=1)
> > 
> >     def forward(self, input_ids_query, input_ids_passage, labels=None, scores=None):
> >         word_embeddings_query = self.word_embedding(input_ids_query)
> >         word_embeddings_passage = self.word_embedding(input_ids_passage)
> > 
> >         query_vector, query_h_n = self.queryRNN(word_embeddings_query)
> >         passage_vector, passage_h_n = self.passageRNN(word_embeddings_passage)
> > 
> >         query_atten_vector = self.queryMatchAtten(query_vector, passage_vector)
> >         passage_atten_vector = self.passageMatchAtten(passage_vector, query_vector)
> > 
> >         final_query_vector = self.queryMaxPooling(query_atten_vector)
> >         final_passage_vector = self.passageMaxPooling(passage_atten_vector)
> > 
> >         final_query_vector = final_query_vector[:,-1,:]
> >         final_passage_vector = final_passage_vector[:,-1,:]
> > 
> >         mul_vector = torch.mul(final_query_vector, final_passage_vector)
> >         tmp = torch.cat((final_query_vector, mul_vector), -1)
> >         query_passage_vector = torch.cat((tmp, final_passage_vector), -1)
> > 
> >         mlp_vector1 = self.mlp1(query_passage_vector)
> >         batch_norm_vector = self.batchNorm(mlp_vector1)
> >         relu_vector = self.relu(batch_norm_vector)
> > 
> >         tmp_logits = self.score_mlp2(relu_vector)
> >         score_logits = torch.sigmoid(tmp_logits)
> >         return score_logits
> > ```
> 
> Hi @xybai,
> 
> Thanks for the reply.
> 
> This is great, but I do need the .onnx file. This will help me to
> 
> 1. Run it through the [onnx model checker](https://github.com/onnx/onnx/blob/master/docs/PythonAPIOverview.md) to determine if it is a valid ONNX model. There could be an exporter bug that could have created an invalid ONNX model which needs to be tracked differently.
> 2. Use the model to actually to score a dummy tensor and debug issues along the way.
> 
> I am not sure the model skeleton will help me do the above. So, if you can share your model, that would be great!
> 
> Thanks!

Hi,

**Comment by xybai** [comment]
> > > > Hi @xybai,
> > > > Could you please share the model for further investigation ?
> > > > Thanks!
> > > 
> > > 
> > > Hi, this is my model skeleton. Please have a look. Thanks!
> > > ```
> > > class MatchAttention(nn.Module):
> > >     def __init__(self):
> > >         super(MatchAttention, self).__init__()
> > >         self.softmax = nn.Softmax(dim=-1)
> > > 
> > >     def forward(self, x, y):
> > >         multi = x.bmm(y.transpose(2, 1))
> > >         weight = self.softmax(multi)
> > >         x_atten = weight.bmm(y)
> > >         return x_atten
> > > 
> > > 
> > > class TestModel(nn.Module):
> > >     def __init__(self):
> > >         super(TestModel, self).__init__()
> > >         self.word_embedding = nn.Embedding(num_embeddings=10000, embedding_dim=100)
> > > 
> > >         self.queryRNN = nn.GRU(input_size=100, hidden_size=256, num_layers=2, batch_first=True,
> > >                                dropout=0.5, bidirectional=False)
> > >         self.passageRNN = nn.GRU(input_size=100, hidden_size=256, num_layers=2, batch_first=True,
> > >                                   dropout=0.5, bidirectional=False)
> > > 
> > >         self.queryMatchAtten = MatchAttention()
> > >         self.passageMatchAtten = MatchAttention()
> > >         self.queryMaxPooling = nn.MaxPool2d(kernel_size=(15, 1), stride=1)
> > >         self.passageMaxPooling = nn.MaxPool2d(kernel_size=(70, 1), stride=1)
> > > 
> > >         self.mlp1 = nn.Linear(in_features=768, out_features=512)
> > >         self.batchNorm = nn.BatchNorm1d(512)
> > >         self.relu = nn.ReLU()
> > >         self.mlp2 = nn.Linear(in_features=512, out_features=2)
> > >         self.score_mlp2 = nn.Linear(in_features=512, out_features=1)
> > > 
> > >     def forward(self, input_ids_query, input_ids_passage, labels=None, scores=None):
> > >         word_embeddings_query = self.word_embedding(input_ids_query)
> > >         word_embeddings_passage = self.word_embedding(input_ids_passage)
> > > 
> > >         query_vector, query_h_n = self.queryRNN(word_embeddings_query)
> > >         passage_vector, passage_h_n = self.passageRNN(word_embeddings_passage)
> > > 
> > >         query_atten_vector = self.queryMatchAtten(query_vector, passage_vector)
> > >         passage_atten_vector = self.passageMatchAtten(passage_vector, query_vector)
> > > 
> > >         final_query_vector = self.queryMaxPooling(query_atten_vector)
> > >         final_passage_vector = self.passageMaxPooling(passage_atten_vector)
> > > 
> > >         final_query_vector = final_query_vector[:,-1,:]
> > >         final_passage_vector = final_passage_vector[:,-1,:]
> > > 
> > >         mul_vector = torch.mul(final_query_vector, final_passage_vector)
> > >         tmp = torch.cat((final_query_vector, mul_vector), -1)
> > >         query_passage_vector = torch.cat((tmp, final_passage_vector), -1)
> > > 
> > >         mlp_vector1 = self.mlp1(query_passage_vector)
> > >         batch_norm_vector = self.batchNorm(mlp_vector1)
> > >         relu_vector = self.relu(batch_norm_vector)
> > > 
> > >         tmp_logits = self.score_mlp2(relu_vector)
> > >         score_logits = torch.sigmoid(tmp_logits)
> > >         return score_logits
> > > ```
> > 
> > 
> > Hi @xybai,
> > Thanks for the reply.
> > This is great, but I do need the .onnx file. This will help me to
> > 
> > 1. Run it through the [onnx model checker](https://github.com/onnx/onnx/blob/master/docs/PythonAPIOverview.md) to determine if it is a valid ONNX model. There could be an exporter bug that could have created an invalid ONNX model which needs to be tracked differently.
> > 2. Use the model to actually to score a dummy tensor and debug issues along the way.
> > 
> > I am not sure the model skeleton will help me do the above. So, if you can share your model, that would be great!
> > Thanks!
> 
> Hi,

Hi @hariharans29 , after talking with one colleague in onnxruntime team. The conversion error maybe related with maxpooling op, while it need 4 dimension, and only 3 is passed into. They will keep track of and try to fix this bug. Many thanks for your time. Thanks!
